src/zos-io.cc

```diff
@@ -854,33 +965,63 @@ FILE *__fopen_ascii(const char *filename, const char *mode) {
   FILE* fp = __fopen_orig(filename, mode);
   int old_errno = errno;
 
+  struct file_tag *t = &sb.st_tag;
+
   if (fp) {
     int fd = fileno(fp);
+    filename = __realpath_extended(filename, NULL);
+    // every RDONLY open we (re)determine filetags
+    if (S_ISREG(sb.st_mode) && (strcmp(mode, "r") == 0)) {
+     existing_files.erase(filename);
+     // if file ccsid doesn't match new file default or if it has the txtflag off
+     // it should be cached, except when it is FT_BINARY which always has the txtflag off
+     if (t->ft_ccsid != __ccsid_new_file() ||
+         !t->ft_ccsid == FT_BINARY && !t->ft_txtflag) {
+        if (t->ft_txtflag)
+           existing_files.insert(std::make_pair(filename, std::make_pair(t->ft_ccsid, true)));
+        else
+           existing_files.insert(std::make_pair(filename, std::make_pair(t->ft_ccsid, false)));
+       }
+    }
+
     if (is_new_file) {
-     __tag_new_file(fd);
+     __tag_new_file_filename(fd, filename);
      errno = old_errno;
     }
     // Enable auto-conversion of untagged files
     else if (S_ISREG(sb.st_mode)) {
       struct file_tag *t = &sb.st_tag;
-      if (t->ft_txtflag == 0 && (t->ft_ccsid == 0 || t->ft_ccsid == 1047) &&
-          strcmp(mode, "r") == 0) {
+      if (t->ft_ccsid == FT_UNTAGGED && strcmp(mode, "r") == 0) {
         if (__file_needs_conversion_init(filename, fd)) {
-          struct f_cnvrt cvtreq = {SETCVTON, 0, 1047};
-          fcntl(fd, F_CONTROL_CVT, &cvtreq);
+          __enableautocvt(fd);
           /* Calling fcntl() should not clobber errno. */
           errno = old_errno;
+          auto it = existing_files.find(filename);
+          if(it != existing_files.end()) {
+             // keep file ccsid 0, but turn on txt conversion
+             // this combination does not exist in 'real life' but it allows
+             // us to distinguise between unconverted and converted untagged files
+             it->second = std::make_pair(FT_UNTAGGED, true);
+          }
         } else {
           // fopen tags untagged files, which enables auto-conversion
           __disableautocvt(fd);
         }
       }
+      else if (t->ft_ccsid == FT_UNTAGGED && strcmp(mode, "r") != 0) {
+          auto it = existing_files.find(filename);
+          if(it != existing_files.end())
+             // if untagged with txtflag off, we have to disable conversion
+             // fopen enables conversion for untagged files by default
+             if(!it->second.second) {
```

Comment:
i am not sure what first and second mean - is there a better name that could be used?

Reply:
Not that I'm aware, it is the way iterators in C++ refer to the first and second 'element' of the std::unordered_maps and the std::pair. In this case the second 'element' in the unordered_maps is an std::pair with ccsid and txtflag. So it->second referes to the std::pair and in that pair 'first' points to the ccsid and 'second' points to the txtflag.